dev/src/write-batch.ts

```diff
@@ -577,7 +586,54 @@ export class WriteBatch implements firestore.WriteBatch {
       api.BatchWriteResponse
     >('batchWrite', request, tag, retryCodes);
 
-    return (response.writeResults || []).map((result, i) => {
+    let retryCount = 0;
+    while (retryCount < MAX_BATCH_WRITE_RETRY_ATTEMPTS) {
+      // Find the indexes of all writes that failed with ABORTED.
+      const abortedIndexes = response.status.reduce(
+        (arr: number[], status, i) => {
+          if (status.code === Status.ABORTED) {
+            arr.push(i);
+          }
+          return arr;
+        },
+        []
+      );
+
+      if (abortedIndexes.length === 0) {
+        break;
+      }
+      logger(
+        'WriteBatch.bulkCommit',
+        tag,
+        'Current batch failed at retry #' +
+          retryCount +
+          '. Num failures: ' +
+          abortedIndexes.length +
+          '/' +
+          response.status.length
+      );
+
+      // Retry the failed writes in a new request.
+      const retryRequest = {...request};
+      retryRequest.writes = retryRequest.writes?.filter((write, i) => {
+        return abortedIndexes.includes(i);
+      });
+      const retriedResponse = await this._firestore.request<
+        api.IBatchWriteRequest,
+        api.BatchWriteResponse
+      >('batchWrite', retryRequest, tag, retryCodes);
+
+      // Map the results of the retried request back to the original response.
+      for (let i = 0; i < abortedIndexes.length; i++) {
+        const originalIndex = abortedIndexes[i];
+        response.writeResults[originalIndex] = retriedResponse.writeResults[i];
+        response.status[originalIndex] = retriedResponse.status[i];
+      }
+      await this._backoff.backoffAndWait();
```

Comment:
The first `backoffAndWait()` doesn't actually do any backoff. Since you call it only after you have already send the first retry this essentially means that only your third attempt will be executed with backoff.